lib/puppet/http/external_client.rb

```diff
@@ -0,0 +1,106 @@
+#
+# Adapts an external http_client_class to the HTTP client API. The former
+# is typically registered by puppetserver and only implements a subset of
+# the Puppet::Network::HTTP::Connection methods. As a result, only the
+# `get` and `post` methods are supported. Calling `delete`, etc will
+# raise a NotImplementedError.
+#
+# @api private
+class Puppet::HTTP::ExternalClient < Puppet::HTTP::Client
+  # Create an external http client
+  #
+  # @param [Class] http_client_class The class to create to handle the request
+  # @api private
+  def initialize(http_client_class)
+    @http_client_class = http_client_class
+  end
+
+  # (see Puppet::HTTP::Client#get)
+  # @api private
+  def get(url, headers: {}, params: {}, options: {}, &block)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
+
+    options[:use_ssl] = url.scheme == 'https'
+    if options[:user] && options[:password]
+      options[:basic_auth] = { user: options[:user], password: options[:password] }
+    end
+
+    client = @http_client_class.new(url.host, url.port, options)
+    response = Puppet::HTTP::Response.new(client.get(url.request_uri, headers, options), url)
+
+    if block_given?
+      yield response
+    else
+      response
+    end
+  rescue Puppet::HTTP::HTTPError
+    raise
+  rescue => e
+    raise Puppet::HTTP::HTTPError.new(e.message, e)
+  end
+
+  # (see Puppet::HTTP::Client#post)
+  # @api private
+  def post(url, headers: {}, params: {}, options: {}, &block)
+    query = encode_params(params)
+    unless query.empty?
+      url = url.dup
+      url.query = query
+    end
+
+    options[:use_ssl] = url.scheme == 'https'
+    if options[:user] && options[:password]
+      options[:basic_auth] = { user: options[:user], password: options[:password] }
+    end
+
+    body = options.fetch(:body) { |_| raise ArgumentError, "'post' requires a 'body' option" }
+    headers['Content-Type'] = options[:content_type]
+    headers['Content-Length'] = body ? body.bytesize.to_s : "0"
+
+    client = @http_client_class.new(url.host, url.port, options)
+    response = Puppet::HTTP::Response.new(client.post(url.request_uri, body, headers, options), url)
+
+    if block_given?
+      yield response
+    else
+      response
+    end
+  rescue Puppet::HTTP::HTTPError
+    raise
+  rescue => e
+    raise Puppet::HTTP::HTTPError.new(e.message, e)
+  end
+
+  # (see Puppet::HTTP::Client#close)
+  # @api private
+  def close
+    # noop
```

Comment:
why is this a noop rather than raise an error? Is it because in this client instance, there is no pool to close? Referring to the docs for Puppet::HTTP::Client#close is a little misleading in that case

Reply:
Updated